Label control implementation

HackerDelphi opened the pull request:
## What does the pull request do?
Adds Label control similar to WPF's one


## What is the current behavior?
Have Label control in Avalonia. 
This can help to migrate WPF XAML files easier than before - Label is activelly used in WPF


## What is the updated/expected behavior with this PR?
Added Label control with Target property


## How was the solution implemented (if it's not obvious)?
Label control is inherited from ContentControl (like in WPF) and adds additional processing of Content to change ContentTemplate if no one is provided to use AccessText for hot key processing..


## Checklist
- [ ] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ x] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation


## Breaking changes
None

## Fixed issues
None

workgroupengineering committed:
Allow support at CanExecute when binding a method

workgroupengineering committed:
Add testing

workgroupengineering committed:
Update BindingDemo Sample

workgroupengineering committed:
Implemented WeakPropertyChangedProxy to avoid memory leak

workgroupengineering committed:
Add test to check ViewModel is Collected

jp2masa committed:
Fire Tapped only on left button click.

Commit:
Expose Cut Copy and Paste methods for binding for context menus

Commit:
add context menu

Commit:
add gesture text.

Commit:
move standard textbox context menu to the templates.

Commit:
revert checkbox page.

Commit:
tidy code.

jkoritzinsky committed:
Re-enable SharpGenTools.Sdk to patch calli but disable codegen (#4512)

kiminuo committed:
AppBuilderBase: Allow to specify app factory.

grokys committed:
Reverted SelectionModel.

Commit:
Fix Missing CaretBrush on DefaultTheme's TextBox

grokys committed:
Fixed/skipped failing tests.

Commit:
Merge pull request #4515 from AvaloniaUI/fixes/fix-caret-default-theme-textbox

Fix Missing CaretBrush on DefaultTheme's TextBox

grokys committed:
Added skipped tests for #4496.

VisualMelon committed:
Fix Issue 3825

Commit:
Merge branch 'master' into feature/2020-08-19-DI

jp2masa committed:
Relaxed conditions for data context type inference in compiled bindings.

maxkatz6 committed:
DataGrid: implement missed pseudoclasses

kiminuo committed:
Address feedback.

Commit:
!F Adding margin/padding

Gillibald committed:
Make sure to always return a valid CharacterHit for next / previous CharacterHit

Gillibald committed:
Fix #4521

Commit:
Merge branch 'master' into feature-devtools-padding-margin

Gillibald committed:
Merge branch 'master' into fixes/TextLineCharacterHit

Gillibald committed:
Merge branch 'master' into fixes/RoundedCorners

Gillibald committed:
Merge pull request #4525 from Gillibald/fixes/RoundedCorners

Fix #4521

pr8x committed:
Merge branch 'master' into feature-devtools-padding-margin

grokys committed:
Reimplemented SelectionModel.

Handles only list selections, not nested selections.

Gillibald committed:
Merge branch 'master' into Issue3825

Commit:
Merge branch 'master' into compiled-bindings-fixes

Commit:
[TextBox] keep caret index when clearing selection.

Commit:
extract clear selection to method.

Commit:
add failing unit test.

Commit:
add context menu to textbox template

Commit:
added context menu to default theme

Commit:
allow osx to run in vm where opengl init may fail.

Commit:
dont crash when window is closed and apis are still called.

Gillibald committed:
Merge branch 'master' into fixes/TextLineCharacterHit

Commit:
Adding support for regex filter of devtools properties

sdoroff committed:
DataGridRowDetails update height

Updates the height of a DataGridRowDetails when its
content changes size

sdoroff committed:
Merge branch 'master' into datagrid-details-height

Commit:
Reduce overhead of compiling regex

Use INotifyDataErrorInfo to signal error

grokys committed:
Added #nullable to SelectingItemsControl.

Commit:
Merge pull request #4528 from AvaloniaUI/fixes/crash-osx-closing-app-when-tooltip-onscreen

dont crash when window is closed and apis are still called.

Commit:
Merge branch 'master' into fixes/allow-osx-to-run-in-vm

Commit:
Merge pull request #4527 from AvaloniaUI/fixes/allow-osx-to-run-in-vm

allow osx to run in vm where opengl init may fail.

Commit:
Merge remote-tracking branch 'origin/master' into fixes/textbox-caret-index-lost-on-focus-lost

Commit:
make clear selection public.

Commit:
add comments and move public method above private and protected.

Commit:
Merge branch 'master' into compiled-bindings-fixes

Commit:
Merge pull request #4519 from jp2masa/compiled-bindings-fixes

Compiled bindings: relaxed conditions for data context type inference

Commit:
Merge branch 'master' into Issue3825

Commit:
Merge pull request #4517 from VisualMelon/Issue3825

Fix Issue3825

Commit:
Merge branch 'master' into fixes/TextLineCharacterHit

Commit:
Merge pull request #4524 from Gillibald/fixes/TextLineCharacterHit

Fix TextLine GetPrevious/NextCharacterHit

Commit:
Merge branch 'master' into feature-devtools-padding-margin

grokys committed:
Started integrating new SelectionModel.

Commit:
Merge branch 'master' into fixes/textbox-caret-index-lost-on-focus-lost

Commit:
Merge pull request #4523 from pr8x/feature-devtools-padding-margin

Adding margin/padding visualization to DevTools

Commit:
Merge pull request #4526 from AvaloniaUI/fixes/textbox-caret-index-lost-on-focus-lost

Fixes/textbox caret index lost on focus lost

Commit:
remove default textboxcontextmenu resource.

grokys committed:
Refactored ItemSourceView.

- Inner list needs to be a non-generic `IList` because `IList<T>` is no covariant.
- Moved out of the Repeater directory as it's now general-purpose.

Commit:
Merge branch 'master' into feature/2020-08-19-DI

Commit:
update xamlil

Commit:
Revert "add context menu to textbox template"

This reverts commit e5085fc9fee833cb3f1961232234a08d86035e0f.

# Conflicts:
#	src/Avalonia.Themes.Fluent/TextBox.xaml

worldbeater committed:
Add .Events projects

Commit:
fix test errors.

jp2masa committed:
Fixed TreeDataTemplate.ItemsSource for compiled bindings.

Commit:
add a contextmenu to resources.

worldbeater committed:
Add Avalonia.ReactiveUI.Events based on Pharmacist

jp2masa committed:
Fixed test.

Commit:
Merge pull request #4514 from kiminuo/feature/2020-08-19-DI

Passing dependencies to Application via constructor.

Commit:
fix mem leak tests for context menus.

grokys committed:
More integration of new SelectionModel.

Commit:
fix memory leak tests.

Commit:
fix leak test.

Commit:
fix another leak test.

grokys committed:
Implemented SelectionMode.AlwaysSelected.

Commit:
add context menu back.

Commit:
Merge remote-tracking branch 'origin/master' into feature/textbox-context-menu

Commit:
restore submodule.

Commit:
cant cut when readonly..

maxkatz6 committed:
Fluent DataGrid styles

worldbeater committed:
Merge branch 'master' into rxui-events

Commit:
use static resource for the context menus in textbox templates.

Commit:
add names for default context menu.

Commit:
Merge pull request #4531 from AvaloniaUI/feature/textbox-context-menu

Feature/textbox context menu

worldbeater committed:
Use latest SDK and runtime

worldbeater committed:
Apply DotNetCoreInstaller workaround

worldbeater committed:
Try using .NET Core 3.1.302

worldbeater committed:
Merge branch 'master' into rxui-events

jkoritzinsky committed:
Merge branch 'master' into features/Binding_Method_CanExecute

worldbeater committed:
Add a smoke test

jkoritzinsky committed:
Merge pull request #4095 from workgroupengineering/features/Binding_Method_CanExecute

[Feature] binding method allow support can execute

worldbeater committed:
Merge branch 'master' into rxui-events

maxkatz6 committed:
Fix datagrid columns sort/resize/reorder #4476

maxkatz6 committed:
DataGrid: do not hardcode padding

maxkatz6 committed:
DataGrid: fix row headers (only fluent)

Commit:
Disable commands when Textbox is on password textbox mode.

Commit:
prevent popup crashing when mainwindow has closed and popup still open.

Commit:
Merge pull request #4545 from maxkatz6/fixes/fix-datagrid-pointer-events

Fix datagrid columns sort/resize/reorder

Commit:
Merge pull request #4547 from AvaloniaUI/fixes/textbox-disable-commands-when-pwdbox

Fixes/textbox disable commands when pwdbox

Commit:
fix expander as not easy to click on, and glyph not visible on fluent theme.

jmacato committed:
Merge pull request #4549 from AvaloniaUI/fixes/expander-glyph

fix expander as not easy to click on, and glyph not visible on fluent…

Commit:
dont make native handle null so closing events can complete.

worldbeater committed:
Merge branch 'master' into rxui-events

maxkatz6 committed:
Merge branch 'master' into feature/fluent-datagrid

maxkatz6 committed:
DataGrid: update samples page

maxkatz6 committed:
DataGrid: support hideable scrollbars and some another fixes

maxkatz6 committed:
DataGrid: rename fluent datagrid resources

grokys committed:
Use CollectionChangedEventManager in ItemsControl.

grokys committed:
Auto-select when items added.

Commit:
prevent screens api crashing if it has been disposed.

Commit:
Merge branch 'master' into fixes/osx-tooltip-crash-on-mainwindow-close

grokys committed:
Copy listeners before notifying.

As raising the event can cause the listeners to change.

grokys committed:
Reimplement commented-out code.

grokys committed:
Adjust a few expected test results.

grokys committed:
Don't raise LostSelection when changing source.

Commit:
ScrollContentPresenter, only block offset property updates during setting of offsetproperty inside arrange.

grokys committed:
Don't autoselect with no items.

grokys committed:
Add missing word.

grokys committed:
Added back removed tests.

grokys committed:
Unskip tests we can handle now.

grokys committed:
Make restored tests pass.

Commit:
Use Array.Empty instead of list.

Co-authored-by: jp2masa <jp2masa@users.noreply.github.com>

Commit:
Merge pull request #4548 from AvaloniaUI/fixes/osx-tooltip-crash-on-mainwindow-close

prevent popup crashing when mainwindow has closed and popup still open.

Commit:
Merge branch 'master' into fixes/fix-scroll-content-presenter

grokys committed:
Use SelectionModel in samples.

grokys committed:
Use SelectionModel in samples.

grokys committed:
Use SelectedIndex in ComboBox.

grokys committed:
Refactor SelectedItemsSync.

Commit:
potential fix for gpu texture leak.

jmacato committed:
Merge pull request #4553 from AvaloniaUI/fixes/fix-scroll-content-presenter

Fix ScrollContentPresenter not Arranging in certain conditions.

Commit:
Merge branch 'master' into fixes/gpu-memory-spike-leak

grokys committed:
Restore selection on reset.

grokys committed:
Adjust expected output.

Commit:
Merge branch 'master' into feature/fluent-datagrid

Commit:
Merge pull request #4534 from maxkatz6/feature/fluent-datagrid

Feature/fluent datagrid

Commit:
Merge remote-tracking branch 'origin/master' into fixes/gpu-memory-spike-leak

Commit:
update to latest skia preview on nuget.

Commit:
update linux native package.

Commit:
Merge branch 'master' into compiled-bindings-fixes

Commit:
Merge pull request #4532 from jp2masa/compiled-bindings-fixes

Fixed TreeDataTemplate.ItemsSource for compiled bindings

grokys committed:
Make remaining tests pass.

grokys committed:
Fix working with collections of value types.

`IEnumerable<T>` is not covariant for value types, so we need to use the non-generic `IEnumerable` everywhere under the hood to ensure we can work with both value and reference types.

grokys committed:
Fix bad grammar.

Ace4896 committed:
Win32: Fix Title for Native OpenFolderDialog

grokys committed:
Implement changing selection mode.

grokys committed:
Commit API compat baseline.

grokys committed:
Merge branch 'master' into refactor/selectionmodel-rewrite

grokys committed:
Merge branch 'master' into datagrid-details-height

Commit:
Merge pull request #4561 from Ace4896/folder-dialog-title

Win32: Fix Title for Native OpenFolderDialog

Commit:
Merge branch 'master' into refactor/selectionmodel-rewrite

Commit:
Merge pull request #4533 from AvaloniaUI/refactor/selectionmodel-rewrite

Rewrite SelectionModel

Commit:
Merge branch 'master' into fixes/gpu-memory-spike-leak

Commit:
add a failing unit test.

Commit:
update _popup.Placement target everytime the context menu is opened.

Commit:
use uithread render timer on windows. to prevent opengl crashing.

Commit:
Merge pull request #4566 from AvaloniaUI/fixes/ui-thread-rendertimer-windows

use uithread render timer on windows. to prevent opengl crashing.

Commit:
Merge pull request #4565 from AvaloniaUI/fixes/crashing-context-menu-when-previously-opened-on-control-no-longer-attached-visual-tree

Fixes Context Menu crashing when opened on a second control and the first no longer in visual tree.

Commit:
Merge branch 'master' into fixes/gpu-memory-spike-leak

Commit:
add styles for password reveal in default theme.

Commit:
use correct case for class names.

Commit:
add clear box to default theme.

jmacato committed:
Merge pull request #4567 from AvaloniaUI/features/password-reveal-styles-default-theme

Add password reveal and Clear content styles to default theme

Commit:
Merge branch 'master' into fixes/gpu-memory-spike-leak

Commit:
Merge pull request #4554 from AvaloniaUI/fixes/gpu-memory-spike-leak

potential fix for gpu texture leak.

Commit:
revert using uithread render timer.

grokys committed:
Merge branch 'master' into datagrid-details-height

worldbeater committed:
Merge branch 'master' into rxui-events

grokys committed:
Merge pull request #4530 from sdoroff/datagrid-details-height

DataGridRowDetails update height on content size change

grokys committed:
Merge branch 'master' into tapped-left-button

grokys committed:
Merge pull request #4494 from arthrp/master

Changing http links to https

grokys committed:
Merge branch 'master' into tapped-left-button

grokys committed:
Merge pull request #4504 from jp2masa/tapped-left-button

Fire Tapped only on left button click

grokys committed:
Don't display vertical grid lines by default.

grokys committed:
Prevent duplicate properties in DevTools.

Properties can be registered on a control as both normal properties and attached properties.

Commit:
Merge branch 'master' into feature-devtool-search-regex

# Conflicts:
#	src/Avalonia.Diagnostics/Diagnostics/ViewModels/TreePageViewModel.cs

Commit:
Compile regex

Commit:
add failing unit test.

Commit:
update tooltip content when already opened and tip changes.

worldbeater committed:
Merge branch 'master' into rxui-events

grokys committed:
Added tests for CollectionChangedEventManager.

Including a failing test: `Receives_Events_From_Wrapped_Collection`.

Commit:
just check newvalue isnt tooltip.

Commit:
Merge remote-tracking branch 'origin/master' into fixes/update-tooltip-content-when-already-opened

Commit:
Merge pull request #4573 from AvaloniaUI/fixes/devtools-duplicate-properties

Prevent duplicate properties in DevTools.

Commit:
Merge pull request #4574 from AvaloniaUI/fixes/update-tooltip-content-when-already-opened

Fixes/update tooltip content when already opened

Commit:
Merge pull request #4572 from AvaloniaUI/fixes/datagrid-gridlines

Don't display vertical grid lines by default.

Commit:
fix innerL/R contents scrolling with text

Commit:
Merge branch 'master' into feature-devtool-search-regex

Commit:
Merge branch 'master' into fixes/pwdbox-hiding-revealbutton

Commit:
Merge pull request #4576 from AvaloniaUI/fixes/pwdbox-hiding-revealbutton

Fixes/pwdbox hiding revealbutton

grokys committed:
Don't use sender as the lookup key.

#4577 was caused because the `sender` of the `CollectionChanged` event didn't correspond to the collection that the event listener was registered on. Instead create an `Entry` object for each collection and put the event handler on that.

Commit:
Merge branch 'master' into fixes/devtools-tree-updates

Commit:
Merge pull request #4578 from AvaloniaUI/fixes/devtools-tree-updates

Fix DevTools tree not updating

pr8x committed:
Merge branch 'master' into feature-devtool-search-regex

Commit:
fix window hit testing.

Commit:
Merge branch 'master' into feature/ios-build

kekekeks committed:
Merge pull request #4586 from AvaloniaUI/feature/ios-build

Feature/ios build

kekekeks committed:
New iOS backend

kekekeks committed:
Merge pull request #4585 from AvaloniaUI/feature/ios-rework-2020

New iOS backend

Commit:
Merge pull request #4583 from AvaloniaUI/fixes/window-hit-testing

fix window hit testing.

Gillibald committed:
Ignore invisible characters

Gillibald committed:
Merge branch 'master' into fixes/IgnoreInvisibleCharacters

Commit:
Merge pull request #4591 from Gillibald/fixes/IgnoreInvisibleCharacters

Ignore invisible characters for TextLine caret navigation

grokys committed:
Merge branch 'master' into feature-devtool-search-regex

Gillibald committed:
Merge pull request #4529 from pr8x/feature-devtool-search-regex

DevTools: Adding support for regex filter

grokys committed:
Addeding failing selection tests.

grokys committed:
Try to handle selections made in CollectionChanged.

There's not much we can do about this except not select invalid indexes.

worldbeater committed:
UseDotNet@2

worldbeater committed:
Merge branch 'master' into rxui-events

Commit:
Merge pull request #4593 from AvaloniaUI/fixes/selectionmodel-collectionchanged

Try to handle selection being changed in CollectionChanged handler.

worldbeater committed:
Try using 3.1.401

worldbeater committed:
Merge branch 'master' into rxui-events

Commit:
update harfbuzz

amwx committed:
Fix SplitView pane to work with popups

Commit:
Merge remote-tracking branch 'origin/master' into fixes/ios-build

Commit:
fix emulator for ios.

Commit:
remove purge resource calls.

Commit:
Merge pull request #4595 from amwx/master

Fix SplitView pane auto close to work with popups

Commit:
Merge pull request #4597 from AvaloniaUI/fixes/ios-build

Fixes/ios build

Commit:
OSX fix dev tools ctrl + shift (implement keyevents on modifier key change events.)

Commit:
Merge pull request #4598 from AvaloniaUI/fixes/osx-modifier-key-events

OSX process modifier key events.

worldbeater committed:
Merge branch 'master' into rxui-events

ShadowDancer committed:
Update Visual.cs documentation

`OnAttachedToVisualTree`/`OnDetachedFromVisualTree`  is called only for root visual tree. Keep xmldoc in line with event description.

grokys committed:
Merge pull request #4535 from worldbeater/rxui-events

Add Avalonia.ReactiveUI.Events based on Pharmacist

Commit:
Add option to enable/disable dirty rects

Commit:
Merge branch 'master' into feature-devtools-dirtyrects

Commit:
Also add option to show fps overlay

Commit:
Fix IOSurface leak.

Commit:
Merge pull request #4612 from AvaloniaUI/fixes/osx-io-surface-mem-leak

Fix IOSurface leak.

Commit:
install typescript with npm.

Commit:
skip building web previewer on osx.

Commit:
Merge pull request #4613 from AvaloniaUI/fixes/osx-ci

Fixes/osx ci

Commit:
Merge branch 'master' into feature-devtools-dirtyrects

Commit:
Merge pull request #4609 from pr8x/feature-devtools-dirtyrects

DevTools: Add option to enable/disable dirty rects and fps overlay

grokys committed:
Merge pull request #4604 from ShadowDancer/patch-1

Update Visual.cs documentation

HackerDelphi committed:
Label control implementation

HackerDelphi committed:
1. Added labels to TextBox page - for testing needs.
2. Also added styles for Labels with same classes as TextBoxes
3. Fixed Label so it does not need Avalonia.Markup anymore.
4. Fixed Label default style to have Padding set to 3 like in WPF

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4618) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4618) before we can accept your contribution.<br/><br/>:x: HackerDelphi [sign now](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=4618)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/AvaloniaUI/Avalonia?pullRequest=4618) it.</sub>

kekekeks commented on src/Avalonia.Controls/Label.cs:
```diff
@@ -0,0 +1,102 @@
+        /// Handler for Content property change event
+        /// </summary>
+        /// <param name="obj">new value</param>
+        private void ContentChanged(object obj)
```
@grokys 
It seems that WPF implements support for access keys directly in ContentPresenter. Should we port that behavior or just keep hacks like this one?

HackerDelphi replied:
> Just curious: Why? Label always seemed like redundant control to me.

It is very needed by users who works with keyboard/hotkeys. I know such users and their work contains of entering many documents with complex structure and ability of using hotkeys to travel through inputs is very important for them. 
They hate programs without hotkeys.

kekekeks commented on src/Avalonia.Themes.Default/Label.xaml:
```diff
@@ -0,0 +1,16 @@
+<Style xmlns="https://github.com/avaloniaui" Selector="Label">
```
Template for the fluent theme is missing (you can copy-paste this one)

maxkatz6 replied:
PR targets 0.9.12 for some reason

HackerDelphi replied:
> Template for the fluent theme is missing (you can copy-paste this one)

Where is it? can't find in Avalonia Repo... :( maybe i'm blind???? ;(

danwalmsley replied:
https://github.com/AvaloniaUI/Avalonia/tree/master/src/Avalonia.Themes.Fluent

HackerDelphi replied:
found it. it was added after 0.9.12

FoggyFinder commented:
Just curious: Why? Label always seemed like redundant control to me.

HackerDelphi committed:
Changed implementation so it works right like WPF version (using ContentPresenter.RecognizesAccessKey property).

grokys committed:
Update ScrollBar visibility with local value priority.

Using `SetValue` with `Style` priority caused a new entry to be added to the `IsVisible` priority store each time the value was updated, causing a leak. Setting it with local value priority subtly changes its semantics but hopefully not so that anyone notices.

worldbeater commented:
Probably worth changing the target branch from `AvaloniaUI:release/0.9.12` to `master`.

kekekeks commented:
@FoggyFinder Label is supposed to be attached to a textbox, so it would activate said textbox on click.

Commit:
Merge pull request #4620 from AvaloniaUI/fixes/scrollbar-leak

Fix ScrollBar leak.

HackerDelphi committed:
merged master

HackerDelphi committed:
Added Fluent theme resources for Label control

HackerDelphi committed:
Markup submodule

HackerDelphi commented:
1. Merged master to branch.
2. Added Fluent theme resources.

Gillibald commented:
Please soft reset your branch clean up everything and force push the result

jmacato commented:
Have to close this since it hasn't been touched and fixed for a month. Also we object in adding redundant controls and im afraid that this `Label` control is one of them.
